File: ios/HeadingCore/Sources/HeadingCore/PrecisionProfile.swift
import Foundation

/// SPEC.md §8 `config/precision-profile-v1.json`, typed.
///
/// Every gate in the engine reads a named key from this type; §18.5 forbids comparing a
/// gate against a numeric literal. There is deliberately **no** calibration-state property:
/// `boundCalibrationState` is derived at runtime from a §24 certification lookup
/// (§19.1, failure mode 32), and §8.1 asserts no key matching `/calibrationState/i` exists
/// anywhere in the document.
public struct PrecisionProfile: Decodable, Equatable {
    public let schemaVersion: String
    public let configVersion: String

    public let orientationMaxAgeMs: Int64
    public let orientationInvalidAfterMs: Int64
    public let freshLocationAtStartMaxAgeMs: Int64
    public let locationAtLockMaxAgeMs: Int64
    public let usableLocationMaxAgeMs: Int64
    public let locationJumpRequiresFreshFixKm: Double
    public let declinationEnvelopeProfessionalMaxDeg: Double
    public let declinationEnvelopeUsableMaxDeg: Double

    public let stableWindowMinMs: Int64
    public let acquisitionTimeoutMs: Int64
    public let periodicOrientationRequestedHz: Double
    public let minPeriodicSupportSamples: Int
    public let clHeadingMinSamplesPerStableWindow: Int
    public let minCircularResultantLength: Double
    public let angularSpeedP95MaxDegPerSec: Double
    public let linearAccelerationP95MaxG: Double
    public let circularResidualP95MaxDeg: Double

    public let flatModePitchAbsMaxDeg: Double
    public let flatModeRollAbsMaxDeg: Double
    public let flatFreehandPlacementBound95Deg: Double
    public let wallNormalElevationAbsMaxDeg: Double
    public let wallTopAxisFromVerticalMaxDeg: Double
    public let wallFreehandPlacementBound95Deg: Double

    public let targetNearZoneDeg: Double
    public let targetCenteringToleranceDeg: Double

    public let providerCrossCheckMaxDeg: Double
    public let referenceSeparationMarginDeg: Double
    public let smallDeclinationAmbiguityMaxDeg: Double
    public let transformAgreementMaxDeg: Double

    public let magneticMagnitudeResidualSuspectFraction: Double
    public let magneticMagnitudeResidualDisturbedFraction: Double
    public let inclinationResidualSuspectDeg: Double
    public let inclinationResidualDisturbedDeg: Double
    public let stationaryFieldMadSuspectMicroTesla: Double
    public let stationaryFieldMadDisturbedMicroTesla: Double
    public let pipelineDisagreementSuspectDeg: Double
    public let pipelineDisagreementDisturbedDeg: Double
    public let suspectInterferenceBound95Deg: Double
    public let recoveryCleanWindowMs: Int64

    public let minHorizontalIntensityNanoTesla: Double

    public let unknownDeviceFloor95Deg: Double
    public let professionalBound95MaxDeg: Double
    public let highBound95MaxDeg: Double
    public let usableBound95MaxDeg: Double
    public let lowConfidenceBound95MaxDeg: Double

    public let spaceWeatherAdvisoryKpMin: Double
    public let spaceWeatherProfessionalSuppressKpMin: Double
    public let spaceWeatherRejectKpMin: Double
    public let spaceWeatherCacheMaxAgeMs: Int64

    public let thermalRestrictionBlocksLock: Bool
    public let wirelessChargingBlocksGradeAboveUsable: Bool

    public let precisionScreenOrientation: String
    public let requireBoundaryStraddleReporting: Bool
    public let geomagneticModelId: String
    public let canonicalAltitudeReference: String
    public let declinationSigmaToBound95Factor: Double

    public static func load(contentsOf url: URL) throws -> PrecisionProfile {
        try JSONDecoder().decode(PrecisionProfile.self, from: Data(contentsOf: url))
    }

    /// The same document as an untyped tree, for the §8.1 whole-document key scan.
    public static func loadRawTree(contentsOf url: URL) throws -> [String: Any] {
        let object = try JSONSerialization.jsonObject(with: Data(contentsOf: url))
        guard let dictionary = object as? [String: Any] else {
            throw SharedArtifacts.ArtifactError.missingArtifact(url.lastPathComponent)
        }
        return dictionary
    }
}

/// Collects every property name appearing anywhere in `value`, at any nesting depth.
/// §8.1's first invariant is "no key matching `/calibrationState/i` exists **anywhere** in
/// the profile", so a nested object may not smuggle one in either.
public func collectPropertyNames(_ value: Any, into names: inout [String]) {
    if let dictionary = value as? [String: Any] {
        for (key, nested) in dictionary {
            names.append(key)
            collectPropertyNames(nested, into: &names)
        }
    } else if let array = value as? [Any] {
        for element in array { collectPropertyNames(element, into: &names) }
    }
}

File: android/heading-core/src/main/kotlin/com/fengshuicompass/headingcore/config/PrecisionProfile.kt
package com.fengshuicompass.headingcore.config

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonArray
import java.io.File

/**
 * SPEC.md §8 `config/precision-profile-v1.json`, typed.
 *
 * Every gate in the engine reads a named key from this type; §18.5 forbids comparing a
 * gate against a numeric literal. The decoder is strict — an unknown key is an error,
 * mirroring the schema's `"additionalProperties": false` — so a key added to the file
 * without a corresponding field fails the build instead of being ignored.
 *
 * There is deliberately **no** calibration-state property, and §8.1 asserts that no key
 * matching `/calibrationState/i` exists anywhere in the document: `boundCalibrationState`
 * is derived at runtime from a §24 certification lookup (§19.1, failure mode 32).
 */
@Serializable
public data class PrecisionProfile(
    val schemaVersion: String,
    val configVersion: String,

    val orientationMaxAgeMs: Long,
    val orientationInvalidAfterMs: Long,
    val freshLocationAtStartMaxAgeMs: Long,
    val locationAtLockMaxAgeMs: Long,
    val usableLocationMaxAgeMs: Long,
    val locationJumpRequiresFreshFixKm: Double,
    val declinationEnvelopeProfessionalMaxDeg: Double,
    val declinationEnvelopeUsableMaxDeg: Double,

    val stableWindowMinMs: Long,
    val acquisitionTimeoutMs: Long,
    val periodicOrientationRequestedHz: Double,
    val minPeriodicSupportSamples: Int,
    val clHeadingMinSamplesPerStableWindow: Int,
    val minCircularResultantLength: Double,
    val angularSpeedP95MaxDegPerSec: Double,
    val linearAccelerationP95MaxG: Double,
    val circularResidualP95MaxDeg: Double,

    val flatModePitchAbsMaxDeg: Double,
    val flatModeRollAbsMaxDeg: Double,
    val flatFreehandPlacementBound95Deg: Double,
    val wallNormalElevationAbsMaxDeg: Double,
    val wallTopAxisFromVerticalMaxDeg: Double,
    val wallFreehandPlacementBound95Deg: Double,

    val targetNearZoneDeg: Double,
    val targetCenteringToleranceDeg: Double,

    val providerCrossCheckMaxDeg: Double,
    val referenceSeparationMarginDeg: Double,
    val smallDeclinationAmbiguityMaxDeg: Double,
    val transformAgreementMaxDeg: Double,

    val magneticMagnitudeResidualSuspectFraction: Double,
    val magneticMagnitudeResidualDisturbedFraction: Double,
    val inclinationResidualSuspectDeg: Double,
    val inclinationResidualDisturbedDeg: Double,
    val stationaryFieldMadSuspectMicroTesla: Double,
    val stationaryFieldMadDisturbedMicroTesla: Double,
    val pipelineDisagreementSuspectDeg: Double,
    val pipelineDisagreementDisturbedDeg: Double,
    val suspectInterferenceBound95Deg: Double,
    val recoveryCleanWindowMs: Long,

    val minHorizontalIntensityNanoTesla: Double,

    val unknownDeviceFloor95Deg: Double,
    val professionalBound95MaxDeg: Double,
    val highBound95MaxDeg: Double,
    val usableBound95MaxDeg: Double,
    val lowConfidenceBound95MaxDeg: Double,

    val spaceWeatherAdvisoryKpMin: Double,
    val spaceWeatherProfessionalSuppressKpMin: Double,
    val spaceWeatherRejectKpMin: Double,
    val spaceWeatherCacheMaxAgeMs: Long,

    val thermalRestrictionBlocksLock: Boolean,
    val wirelessChargingBlocksGradeAboveUsable: Boolean,

    val precisionScreenOrientation: String,
    val requireBoundaryStraddleReporting: Boolean,
    val geomagneticModelId: String,
    val canonicalAltitudeReference: String,
    val declinationSigmaToBound95Factor: Double,
) {
    public companion object {
        /** Strict: rejects unknown keys, exactly as the schema rejects extra properties. */
        private val strictJson = Json {
            ignoreUnknownKeys = false
            isLenient = false
            allowSpecialFloatingPointValues = false
        }

        public fun load(file: File): PrecisionProfile =
            strictJson.decodeFromString(serializer(), file.readText())

        /** The same document as an untyped tree, for the §8.1 whole-document key scan. */
        public fun loadRawTree(file: File): JsonObject =
            strictJson.parseToJsonElement(file.readText()) as JsonObject
    }
}

/**
 * Collects every property name appearing anywhere in [element], at any nesting depth.
 * §8.1's first invariant is "no key matching `/calibrationState/i` exists **anywhere** in
 * the profile", so a nested object may not smuggle one in either.
 */
public fun collectPropertyNames(element: JsonElement, into: MutableList<String> = mutableListOf()): List<String> {
    when (element) {
        is JsonObject -> element.forEach { (name, value) ->
            into += name
            collectPropertyNames(value, into)
        }
        is JsonArray -> element.forEach { collectPropertyNames(it, into) }
        else -> Unit
    }
    return into
}
